### PR TITLE
[Binary parser] refactoring to make room for more loop types

### PR DESCRIPTION
**Pull request by olgavrou**
- creating more dir structure
- moving everything cb related into separate file (`typed_event.h`) and using template specialization struct to provide basic event processing functionality

next: make multistep joiner use the cb related functions that are the same

### CODE REVIEW

File: external_parser/event_processors/joined_event.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
+#pragma once
```
Comment (olgavrou) [inline external_parser/event_processors/joined_event.h:1]: just moving this stuff from `reward.h` into separate file

File: external_parser/event_processors/reward.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
+#pragma once
```
Comment (olgavrou) [inline external_parser/event_processors/reward.h:1]: just moved file

**Review by lalo — APPROVED** [review]

File: external_parser/unit_tests/test_vw_binary_parser.cc [hunk 3]
```diff
@@ -346,9 +353,13 @@ BOOST_AUTO_TEST_CASE(test_log_file_with_invalid_cb_context) {
...
   }
 
+  std::cout << "total size of examples: " << total_size_of_examples << std::endl;
+
   // no interactions processed
   BOOST_CHECK_EQUAL(total_size_of_examples, 0);
+  std::cout << "clear" << std::endl;
 
   clear_examples(examples, vw);
+  std::cout << "vw finish" << std::endl;
```
Comment (cheng-tan) [inline external_parser/unit_tests/test_vw_binary_parser.cc:363]: should we remove these prints?

File: external_parser/unit_tests/test_common.cc [hunk 4]
```diff
@@ -30,15 +30,20 @@ void set_buffer_as_vw_input(const std::vector<char> &buffer, vw *vw) {
 }
 
 std::vector<char> read_file(std::string file_name) {
-  std::ifstream file(file_name, std::ios::binary | std::ios::ate);
-  std::streamsize size = file.tellg();
+  std::cout << "file name: " << file_name << std::endl;
```
Comment (cheng-tan) [inline external_parser/unit_tests/test_common.cc:33]: should we remove these prints?

**Review by cheng-tan — APPROVED** [review]

File: external_parser/unit_tests/test_vw_binary_parser.cc (hunk 3 above)
Comment (olgavrou) [reply external_parser/unit_tests/test_vw_binary_parser.cc]: yup done

File: external_parser/unit_tests/test_common.cc (hunk 4 above)
Comment (olgavrou) [reply external_parser/unit_tests/test_common.cc]: yes removed

**Pull request merged**